### PR TITLE
Overwrite doc.consumes decorator

### PR DESCRIPTION
**Pull request by canberkkoc**
Changed the decorator required parameter processing style. 
It is a design concern for project. At this moment if you have a usage like this:
image
decorator overwrite the field params as the code have default parameter. So i have changed this to use field params. Also i wonder if you have more convenient way to provide this feature.

### CODE REVIEW

File: sanic_openapi/doc.py [hunk 1]
```diff
@@ -290,7 +290,11 @@ def consumes(*args, content_type=None, location="query", required=False):
     def inner(func):
         if args:
             for arg in args:
-                field = RouteField(arg, location, required)
+                if arg.required:
+                    req = arg.required
+                else:
+                    req = required
+                field = RouteField(arg, location, req)
```
[the remark this answers is not in the document]
Comment (canberkkoc) [reply sanic_openapi/doc.py]: Actually i made this PR for pointing my idea. I'll fix it

[the remark this answers is not in the document]
Comment (canberkkoc) [reply sanic_openapi/doc.py]: BTW with last commit only asserting difference tests failing because doc.consumes decorator changed behaviour.

**Comment by chenjr0719** [comment]
Hi @canberkkoc 
I have read your commit message. Here is my feedback:

```python
@doc.consumes({"query_id": doc.String(required=True),"x":doc.String(required=True)}, location="query")
```

In this case, the actual `arg` is a `dict`. And, `sanic-openapi` will convert this `dict` to a `Dictionary()` field. Here is related code:
https://github.com/huge-success/sanic-openapi/blob/a99b98f03425c38248cb1b0f5f6f21b0a5498d81/sanic_openapi/doc.py#L181-L182

So, in this case, it won't loop over this `dict` and this is expected behavior.

About that `gettar(arg, 'required' ,required)` return `None`, I think it is because of this:
https://github.com/huge-success/sanic-openapi/blob/a99b98f03425c38248cb1b0f5f6f21b0a5498d81/sanic_openapi/doc.py#L5-L6

As you can see, the default value of `required` is `None`, so it might return `None` is some cases. To fix this problem, I think it is acceptable to make it default to `False`.

**Comment by canberkkoc** [comment]
> Hi @canberkkoc
> I have read your commit message. Here is my feedback:
> 
> ```python
> @doc.consumes({"query_id": doc.String(required=True),"x":doc.String(required=True)}, location="query")
> ```
> 
> In this case, the actual `arg` is a `dict`. And, `sanic-openapi` will convert this `dict` to a `Dictionary()` field. Here is related code:
> 
> https://github.com/huge-success/sanic-openapi/blob/a99b98f03425c38248cb1b0f5f6f21b0a5498d81/sanic_openapi/doc.py#L181-L182
> 
> So, in this case, it won't loop over this `dict` and this is expected behavior.
> 
> About that `gettar(arg, 'required' ,required)` return `None`, I think it is because of this:
> 
> https://github.com/huge-success/sanic-openapi/blob/a99b98f03425c38248cb1b0f5f6f21b0a5498d81/sanic_openapi/doc.py#L5-L6
> 
> As you can see, the default value of `required` is `None`, so it might return `None` is some cases. To fix this problem, I think it is acceptable to make it default to `False`.

Hi @chenjr0719 

I got it, changed default required option to False and tried. With last commit consumes work like what i think, tests fail because of behaviour change of decorator. 
Btw if we can write more explanatory,varied docs it would be very useful. Thank you for quick response.

**Comment by chenjr0719** [comment]
@canberkkoc Sorry for lost contact a while. I think it is a little complex of this PR now. I will break this PR into two parts.

1. **Enable use `required` from `arg`**:
In this part, I think you already do a really good work to enable this feature. Thanks a lot.

2. **Override mechanism of `required`**:
This is the complex part of this PR now. Here are some examples and I want to know what do you think about those cases:

    **case 1**:

    ```python
    @app.post("/")
    @doc.consumes(doc.String())
    def test(request):
        return text("test")


    _, response = app.test_client.get("/swagger/swagger.json")
    swagger_json = response.json
    path = swagger_json["paths"]["/"]["post"]
    assert path["parameters"][0]["required"] == False  # By default, it should be False.

    ```

    **case 2**:

    ```python
    @app.post("/")
    @doc.consumes(doc.String(required=True))
    def test(request):
        return text("test")


    _, response = app.test_client.get("/swagger/swagger.json")
    swagger_json = response.json
    path = swagger_json["paths"]["/"]["post"]
    assert path["parameters"][0]["required"] == True  # Because the required in field is True and no required in consumes, this should be True.
    ```

    **case 3**:

    ```python
    @app.post("/")
    @doc.consumes(doc.String(), required=True)
    def test(request):
        return text("test")


    _, response = app.test_client.get("/swagger/swagger.json")
    swagger_json = response.json
    path = swagger_json["paths"]["/"]["post"]
    assert path["parameters"][0]["required"] == True  # Because the required in consumes() is True and no required in field, this should be True.

    ```

    And the problem is coming at next two examples.
    **case 4**:

    ```python
    @app.post("/")
    @doc.consumes(doc.String(required=False), required=True)
    def test(request):
        return text("test")


    _, response = app.test_client.get("/swagger/swagger.json")
    swagger_json = response.json
    path = swagger_json["paths"]["/"]["post"]
    assert path["parameters"][0]["required"] == True  # This should be True because the required in consumes() should override the required in field.

    ```

    **case 5**:

    ```python
    @app.post("/")
    @doc.consumes(doc.String(required=True), required=False)
    def test(request):
        return text("test")


    _, response = app.test_client.get("/swagger/swagger.json")
    swagger_json = response.json
    path = swagger_json["paths"]["/"]["post"]
    assert path["parameters"][0]["required"] == False # This should be Fasle because required in consumes() should override the required in field.

    ```

I think we have to clarify how this two `required` works at the same time or it will be a mess. And I think I have to apologize to you because I might have some misleading before. :worried:

**Comment by canberkkoc** [comment]
> @canberkkoc Sorry for lost contact a while. I think it is a little complex of this PR now. I will break this PR into two parts.
> 
>     1. **Enable use `required` from `arg`**:
>        In this part, I think you already do a really good work to enable this feature. Thanks a lot.
> 
>     2. **Override mechanism of `required`**:
>        This is the complex part of this PR now. Here are some examples and I want to know what do you think about those cases:
>        **case 1**:
>        ```python
>        @app.post("/")
>        @doc.consumes(doc.String())
>        def test(request):
>            return text("test")
>        
>        
>        _, response = app.test_client.get("/swagger/swagger.json")
>        swagger_json = response.json
>        path = swagger_json["paths"]["/"]["post"]
>        assert path["parameters"][0]["required"] == False  # By default, it should be False.
>        ```
>        
>        
>        **case 2**:
>        ```python
>        @app.post("/")
>        @doc.consumes(doc.String(required=True))
>        def test(request):
>            return text("test")
>        
>        
>        _, response = app.test_client.get("/swagger/swagger.json")
>        swagger_json = response.json
>        path = swagger_json["paths"]["/"]["post"]
>        assert path["parameters"][0]["required"] == True  # Because the required in field is True and no required in consumes, this should be True.
>        ```
>        
>        
>        **case 3**:
>        ```python
>        @app.post("/")
>        @doc.consumes(doc.String(), required=True)
>        def test(request):
>            return text("test")
>        
>        
>        _, response = app.test_client.get("/swagger/swagger.json")
>        swagger_json = response.json
>        path = swagger_json["paths"]["/"]["post"]
>        assert path["parameters"][0]["required"] == True  # Because the required in consumes() is True and no required in field, this should be True.
>        ```
>        
>        
>        And the problem is coming at next two examples.
>        **case 4**:
>        ```python
>        @app.post("/")
>        @doc.consumes(doc.String(required=False), required=True)
>        def test(request):
>            return text("test")
>        
>        
>        _, response = app.test_client.get("/swagger/swagger.json")
>        swagger_json = response.json
>        path = swagger_json["paths"]["/"]["post"]
>        assert path["parameters"][0]["required"] == True  # This should be True because the required in consumes() should override the required in field.
>        ```
>        
>        
>        **case 5**:
>        ```python
>        @app.post("/")
>        @doc.consumes(doc.String(required=True), required=False)
>        def test(request):
>            return text("test")
>        
>        
>        _, response = app.test_client.get("/swagger/swagger.json")
>        swagger_json = response.json
>        path = swagger_json["paths"]["/"]["post"]
>        assert path["parameters"][0]["required"] == False # This should be Fasle because required in consumes() should override the required in field.
>        ```
> 
> 
> I think we have to clarify how this two `required` works at the same time or it will be a mess. And I think I have to apologize to you because I might have some misleading before. worried

Hello again i think i must apologize too it is a busy period going on for me. 
When i made this PR i thought about this concerns too. 
I think if we use inner required arg like doc.String(required=True) it should override outer required arg. And we should emphasize this matter on docs.

**Pull request closed without merging**